### ISSUE TITLE
Swiping is janky in mobile Firefox

### ISSUE THREAD

**Issue by nolanlawson** [tootsuite/mastodon#5162]
On Firefox for Android, swiping columns or modal images is very low-framerate and unresponsive, especially compared to Chrome.

- Firefox: https://toot.cafe/media/Izx6uZ689MYp7JBXbdA
- Chrome: https://toot.cafe/media/_ILxmW9uDYWuetGbGDY
- Discussion: https://dev.glitch.social/@ashkitten/913393

Tested in Firefox Nightly 58.0a1, commit 1df453aff6f46e1fddc665bffa3f68312f672956.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by nolanlawson** [comment]
I tried `will-change: transform`, no effect. My hunch is that the issue is in `react-swipeable-views`, specifically [the touchmove logic](https://github.com/oliviertassinari/react-swipeable-views/blob/5835f94383ddbc7d16baea4df8404d3f70262830/packages/react-swipeable-views/src/SwipeableViews.js#L444-L457). I noticed it's neither debouncing nor using `requestAnimationFrame`, so it may be relying on unspecced/vendor-specific Chrome behavior on when exactly the browser fires `mousemove` events. E.g. maybe Firefox fires the events more frequently than Chrome does, or maybe Chrome syncs these events to the frames.

**Comment by sorin-davidoi** [comment]
We *might* be saved by Stylo and WebRender - enabled both in Nightly and it does not jank as much.

**Comment by sorin-davidoi** [comment]
The bottleneck is not the animation per-say, but React - instead of setting the transform property on the DOM from inside the event listener, the library calls `setState` and relies on `render` to update the DOM...

![image](https://user-images.githubusercontent.com/2109702/32245328-bbf6c5a0-be7b-11e7-9fa1-8415b12bd854.png)
![image](https://user-images.githubusercontent.com/2109702/32245394-e9ae6dea-be7b-11e7-8916-8f1db3cfce28.png)

**Comment by sorin-davidoi** [comment]
Adding the following on line 428 (before the call to `setState`) in `node_modules/react-swipeable-views/lib/SwipeableViews.js` makes swiping smooth, but sometimes the animation remains in an intermediary state:

```js
if (_this.state.isDragging === true) {
  // Avoid re-rendering to speed-up the animation
  _this.containerNode.style.transform = axisProperties.transform[_this.props.axis](index * 100);

  if (onSwitching) {
    onSwitching(index, 'move');
  }

  return;
}
```

**Comment by nolanlawson** [comment]
What about throttling/debouncing the `touchmove` event using `requestAnimationFrame()`? It seems like the event is firing faster than Firefox can draw frames, although I'm not sure why it's not an issue in Chrome…

**Comment by sorin-davidoi** [comment]
It doesn't fire too fast - if we don't trigger a React render the frames are drawn just fine. It is just that React takes too much time, blocking the main thread and making the page skip frames.

**Comment by nolanlawson** [comment]
Hmmm so maybe Chrome is JITing those functions whereas SpiderMonkey is not? This may be useful to report to Mozilla as a perf bug, or alternatively to try to move functions around so that they're more likely to get JITed. (E.g. does this code path have a lot of functions-within-functions or something?)

**Comment by nolanlawson** [comment]
This seems fixed in Mastodon 2.1.0 based on my testing, probably thanks to https://github.com/oliviertassinari/react-swipeable-views/pull/372

**Comment by oliviertassinari** [comment]
Great! :)
